File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
bool valid(TreeNode* root, long long mini, long long maxi) {
    if (!root) return true;
    if (root->val >= maxi || root->val <= mini) return false;
    return valid(root->left, mini, root->val) && valid(root->right, root->val, maxi);
}

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return valid(root, LONG_MIN, LONG_MAX);
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public void preorder(TreeNode root, List<Integer> list) {
        if(root == null) return;
        if(root.left != null) preorder(root.left,list);
        list.add(root.val);
        if(root.right != null) preorder(root.right,list);

    }


    public boolean isValidBST(TreeNode root) {
        ArrayList<Integer> list = new ArrayList<>();
        preorder(root,list);
        int n = list.size();
        for(int i=0; i<n-1; i++) {
            if(list.get(i) >= list.get(i+1)) return false;
        }
        return true;
    }
}